src/full/Agda/TypeChecking/Sort.hs

```diff
@@ -85,57 +86,66 @@ hasBiggerSort = void . inferUnivSort
+      return $ PiSort (unEl <$> a) (ignoreBlocking s1') (ignoreBlocking <$> s2')
 
-  --case piSort' s1 s2 of
-  --  Just s -> return s
-  --  Nothing -> do
-  --    addConstraint $ HasPTSRule s1 s2
-  --    return $ PiSort s1 s2
-
-  return $ piSort (unEl <$> a) s1' s2'
 
 -- | As @inferPiSort@, but for a nondependent function type.
 inferFunSort :: Sort -> Sort -> TCM Sort
-inferFunSort s1 s2 = funSort <$> reduce s1 <*> reduce s2
-
-ptsRule :: Dom Type -> Abs Sort -> Sort -> TCM ()
-ptsRule a b c = do
-  c' <- inferPiSort a b
-  ifM (optCumulativity <$> pragmaOptions)
-    (leqSort c' c)
-    (equalSort c' c)
-
--- | Non-dependent version of ptsRule
-ptsRule' :: Sort -> Sort -> Sort -> TCM ()
-ptsRule' a b c = do
-  c' <- inferFunSort a b
-  ifM (optCumulativity <$> pragmaOptions)
-    (leqSort c' c)
-    (equalSort c' c)
+inferFunSort s1 s2 = do
+  ss1 <- sortOf (Sort s1)
+  s1' <- reduceB s1
+  s2' <- reduceB s2
+  case funSort' (ignoreBlocking s1') (ignoreBlocking s2') of
+    Right s -> return s
+    Left b -> do
+      let b' = unblockOnEither (getBlocker s1') (getBlocker s2')
+      addConstraint (unblockOnEither b b') $ HasPTSRule (defaultDom (El (univSort s1) (Sort s1))) (mkAbs "_" $ ignoreBlocking s2')
+      return $ FunSort s1 s2
 
 hasPTSRule :: Dom Type -> Abs Sort -> TCM ()
-hasPTSRule a b = void $ inferPiSort a b
+hasPTSRule a b =
+  if alwaysValidCodomain $ unAbs b
+  then return ()
+  else do
+    infer <- reduceB =<< inferPiSort a b
+    case infer of
+      Blocked b t | neverUnblock == b -> typeError $ InvalidTypeSort t
+      NotBlocked _ t@FunSort{}        -> typeError $ InvalidTypeSort t
+      NotBlocked _ t@PiSort{}         -> typeError $ InvalidTypeSort t
+      _ -> return ()
+  where
+    alwaysValidCodomain s | Right LargeSort{} <- sizeOfSort s
+                               = True
+    alwaysValidCodomain Type{} = True
+    alwaysValidCodomain Prop{} = True
+    alwaysValidCodomain SSet{} = True
+    alwaysValidCodomain (PiSort _ _ s) = alwaysValidCodomain $ unAbs s
+    alwaysValidCodomain (FunSort _ s) = alwaysValidCodomain s
+    alwaysValidCodomain _      = False
 
 -- | Recursively check that an iterated function type constructed by @telePi@
 --   is well-sorted.
 checkTelePiSort :: Type -> TCM ()
 -- Jesper, 2019-07-27: This is currently doing nothing (see comment in inferPiSort)
---checkTelePiSort (El s (Pi a b)) = do
---  -- Since the function type is assumed to be constructed by @telePi@,
---  -- we already know that @s == piSort (getSort a) (getSort <$> b)@,
---  -- so we just check that this sort is well-formed.
---  hasPTSRule a (getSort <$> b)
---  underAbstraction a b checkTelePiSort
+checkTelePiSort (El s (Pi a b)) = do
+  -- Since the function type is assumed to be constructed by @telePi@,
+  -- we already know that @s == piSort (getSort a) (getSort <$> b)@,
+  -- so we just check that this sort is well-formed.
+  hasPTSRule a (getSort <$> b)
+  underAbstraction a b checkTelePiSort
```

Comment:
This function was reactivated, so now the comment by @jespercockx is stale ("currently doing nothing").
@arthur-adjedj Please add a new comment justifying why it now has to do something.